src/main/java/org/elasticsearch/action/search/type/TransportSearchDfsQueryAndFetchAction.java

```diff
@@ -105,6 +105,10 @@ public void onResponse(QueryFetchSearchResult result) {
                 @Override
                 public void onFailure(Throwable t) {
                     onSecondPhaseFailure(t, querySearchRequest, shardIndex, dfsResult, counter);
+                    // the query might not have been executed at all (for example because thread pool rejected execution)
+                    // and the search context that was created in dfs phase might not be released.
+                    // release it again to be in the safe side
+                    sendReleaseSearchContext(querySearchRequest.id(), node);
```

Comment:
Even if onSecondPhaseFailure is not supposed to throw exceptions, should we put onSecondPhaseFailure in a try block and sendReleaseSearchContext in the finally to be on the safe side?